### DISCUSSION TITLE
Selection Set seems wrong

### DISCUSSION THREAD

**Discussion by schaefa** [graphql-java/graphql-java#3084] [General]
I have this GraphQL query: 
```
{ 
 combinedFetcher { 
    boolValue 
    interfaceTest { 
       id 
       ... on Human { address } 
       ... on Droid { primaryFunction } 
     } 
 } 
}
```

When I run this through graphql-java 17.4 I see a Selection Set that has 3 fields in the interfaceTest selection set:
- [Droid,Human].id
- Human.address
- Droid.primaryFunction
Both Human and Droid object have an id field.

Does this mean the '.id' field can be in interfaceTest, interfaceTest/Human and interfaceTest/Droid?
In v15 the Human and Droid would be an InlineFragment and so I would know that they are not fields.

**Comment by bbakerman** [comment]
This is expected behavior.  

Since you didnt post your schema lets assume that the `interfaceTest` field type is an interface called `Character`

The `interfaceTest` field is executed and returns some object.  We dont know what at the time the field is invoked

It will resolve (via the type resolver) to an object of type `Human` or `Droid` but at the time the DataFetcher is invoked we don't yet know.

So the selection set is

  * `[Droid,Human].id`
     *  since both types have an `id` then it could be either a `Droid.id` or a `Human.id` in selection terms
     *   we dont know yet what it will be exactly but thats why it has ` [Droid,Human].id`
  * `Human.address`
      *  This is listed because the object  may resolve to be a `Human` and hence the `Human.address` is a possible selection
   * `Droid.primaryFunction`
      *  This is listed because the object  may resolve to be a `Droid` and hence the `Droid.primaryFunction` is a possible selection

> Does this mean the '.id' field can be in interfaceTest, interfaceTest/Human and interfaceTest/Droid?

This means that the id field (shared in the two concrete types `Human` and `Droid`) is present and once the object it reseloved (after data fetching) then it's going to be one of the `id` fields in two alternative types.

Most of the time you dont care.  Its the `id` field and you can just treat it without regard to its containing object type.

But for certain uses cases the precise containing object type and field is important and hence we have "maybe" indications insode the selection set.

The fragments (inline or otherwise) have nothing to do with selection once they are normalised down.  

For example this silly but correct selection is possible

```
{ 
 combinedFetcher { 
    boolValue 
    interfaceTest { 
       id 
       ... on Human { id, address } 
       ... on Droid { id, primaryFunction } 
       ... on Character {
          id 
       }
     } 
 } 
}
```

The above will all be merged down to the one id field in a sub selection sense and its can be of 1 of 2 concrete object types (`Human ` or `Droid`)

Read this blog about how funky field selection can be and how fragments don't really represent the fields to be selected at execution runtime

https://www.graphql-java.com/blog/deep-dive-merged-fields

As for V15 - we had a different and less correct system in place.  It could not show the possible alternatives properly and this was fixed in later versions.

**Comment by schaefa** [comment]
I am not an expert in GraphQL query syntax but I am tasked with upgrading Sling GraphQL Core to 17.4 or higher because of the possible DoS attack.
Because of that I tried first to see what breaks and then resolve it but this test class: https://github.com/apache/sling-org-apache-sling-graphql-core/blob/master/src/test/java/org/apache/sling/graphql/core/engine/DefaultQueryExecutorTest.java yields a breaking change.
So Sling so far used the Selection Set to create a hierarchical structure of the Query but with that change this seems not to be possible anymore. I need to see how this is used and then see how to adjust.
Thanks for the help.
